### ISSUE TITLE
weave-npc crashing on k8s 1.7.4

### ISSUE THREAD

**Issue by bismarck** [weaveworks/weave#3095]
## What you expected to happen?
weave 2.0.2 pods should start without an issue

## What happened?
weave-npc container crashes

## How to reproduce it?

On a k8s 1.7.4 cluster that has some network policies defined run
kubectl apply -f https://cloud.weave.works/k8s/v1.6/net

## Anything else we need to know?

We are running k8s on AWS, built with KOPS.

## Versions:

```
$ weave version
weave script 2.0.2
$ docker version
Client:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:17:57 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.6
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   78d1802
 Built:        Tue Jan 10 20:17:57 2017
 OS/Arch:      linux/amd64
$ uname -a
Linux ip-10-203-20-75 4.4.78-k8s #1 SMP Fri Jul 28 01:28:39 UTC 2017 x86_64 GNU/Linux
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4", GitCommit:"793658f2d7ca7f064d2bdf606519f9fe1229c381", GitTreeState:"clean", BuildDate:"2017-08-17T17:03:51Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.4", GitCommit:"793658f2d7ca7f064d2bdf606519f9fe1229c381", GitTreeState:"clean", BuildDate:"2017-08-17T08:30:51Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

## Logs:
```
$ kubectl logs -n kube-system <weave-net-pod> weave -c weave-npc
INFO: 2017/08/22 18:09:46.888634 Starting Weaveworks NPC 2.0.2; node name "ip-10-203-19-219.ec2.internal"
INFO: 2017/08/22 18:09:46.905166 Serving /metrics on :6781
Tue Aug 22 18:09:46 2017 <5> ulogd.c:843 building new pluginstance stack: 'log1:NFLOG,base1:BASE,pcap1:PCAP'
DEBU: 2017/08/22 18:09:46.921549 Got list of ipsets: [weave-local-pods weave-^{gg_XX+0pde8@K:E1%GeGc2_]
DEBU: 2017/08/22 18:09:46.921594 Flushing ipset 'weave-local-pods'
DEBU: 2017/08/22 18:09:46.922724 Flushing ipset 'weave-^{gg_XX+0pde8@K:E1%GeGc2_'
DEBU: 2017/08/22 18:09:46.923596 Destroying ipset 'weave-local-pods'
DEBU: 2017/08/22 18:09:46.924484 Destroying ipset 'weave-^{gg_XX+0pde8@K:E1%GeGc2_'
INFO: 2017/08/22 18:09:46.961006 EVENT AddNetworkPolicy {"metadata":{"name":"alert-transformer-listener","namespace":"dev-monitoring","selfLink":"/apis/extensions/v1beta1/namespaces/dev-monitoring/networkpolicies/alert-transformer-listener","uid":"c86d8f1b-7dd3-11e7-bf26-066013bc42ce","resourceVersion":"23095942","generation":1,"creationTimestamp":"2017-08-10T13:57:01Z"},"spec":{"podSelector":{"matchLabels":{"app":"alert-transformer-listener"}},"ingress":[{"from":[{"podSelector":{"matchExpressions":[{"key":"app","operator":"In","values":["alert-transformer"]}]}}]},{"ports":[{"protocol":"TCP","port":8080},{"protocol":"UDP","port":9225}]}]}}
INFO: 2017/08/22 18:09:46.961644 creating ipset: &npc.selectorSpec{key:"", selector:labels.internalSelector{}, ipsetType:"hash:ip", ipsetName:"weave-^{gg_XX+0pde8@K:E1%GeGc2_", nsName:"dev-monitoring"}
E0822 18:09:46.964079   27960 runtime.go:64] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:70
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:63
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:49
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:290
/go/src/github.com/weaveworks/weave/npc/rule.go:34
/go/src/github.com/weaveworks/weave/npc/analyser.go:49
/go/src/github.com/weaveworks/weave/npc/analyser.go:112
/go/src/github.com/weaveworks/weave/npc/analyser.go:51
/go/src/github.com/weaveworks/weave/npc/namespace.go:171
/go/src/github.com/weaveworks/weave/npc/controller.go:124
/go/src/github.com/weaveworks/weave/npc/controller.go:76
/go/src/github.com/weaveworks/weave/npc/controller.go:125
/go/src/github.com/weaveworks/weave/prog/weave-npc/main.go:206
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:168
<autogenerated>:50
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:246
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:126
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:102
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:87
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:88
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:49
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:102
/usr/local/go/src/runtime/asm_amd64.s:2197
E0822 18:09:46.965887   27960 runtime.go:64] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:70
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:63
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:49
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:56
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:290
/go/src/github.com/weaveworks/weave/npc/rule.go:34
/go/src/github.com/weaveworks/weave/npc/analyser.go:49
/go/src/github.com/weaveworks/weave/npc/analyser.go:112
/go/src/github.com/weaveworks/weave/npc/analyser.go:51
/go/src/github.com/weaveworks/weave/npc/namespace.go:171
/go/src/github.com/weaveworks/weave/npc/controller.go:124
/go/src/github.com/weaveworks/weave/npc/controller.go:76
/go/src/github.com/weaveworks/weave/npc/controller.go:125
/go/src/github.com/weaveworks/weave/prog/weave-npc/main.go:206
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:168
<autogenerated>:50
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:246
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:126
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:102
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:87
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:88
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:49
/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:102
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x1245c9e]

goroutine 39 [running]:
github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:56 +0x126
panic(0x14cdbe0, 0x2162f60)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:56 +0x126
panic(0x14cdbe0, 0x2162f60)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/weaveworks/weave/npc.newRuleSpec(0xc421549830, 0x0, 0xc4214fb860, 0xc421549840, 0x4)
	/go/src/github.com/weaveworks/weave/npc/rule.go:34 +0x38e
github.com/weaveworks/weave/npc.(*ns).analysePolicy.func1(0xc42151e898, 0x3, 0xc421570e7c, 0x4)
	/go/src/github.com/weaveworks/weave/npc/analyser.go:49 +0x54
github.com/weaveworks/weave/npc.withNormalisedProtoAndPort(0xc421521600, 0x2, 0x2, 0xc421549950)
	/go/src/github.com/weaveworks/weave/npc/analyser.go:112 +0xa3
github.com/weaveworks/weave/npc.(*ns).analysePolicy(0xc42007def0, 0xc421518a80, 0xc421569a70, 0xc4214eb5a8, 0x8, 0x907d86f21431380e, 0x0)
	/go/src/github.com/weaveworks/weave/npc/analyser.go:51 +0x6df
github.com/weaveworks/weave/npc.(*ns).addNetworkPolicy(0xc42007def0, 0xc421518a80, 0xc4214eb450, 0x216b910)
	/go/src/github.com/weaveworks/weave/npc/namespace.go:171 +0x7f
github.com/weaveworks/weave/npc.(*controller).AddNetworkPolicy.func1(0xc42007def0, 0xc4214c8b70, 0xc421549b10)
	/go/src/github.com/weaveworks/weave/npc/controller.go:124 +0x34
github.com/weaveworks/weave/npc.(*controller).withNS(0xc420014fa0, 0xc42151e7a0, 0xe, 0xc421549b60, 0x1, 0x1)
	/go/src/github.com/weaveworks/weave/npc/controller.go:76 +0x95
github.com/weaveworks/weave/npc.(*controller).AddNetworkPolicy(0xc420014fa0, 0xc421518a80, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/npc/controller.go:125 +0x16d
main.root.func7(0x163c220, 0xc421518a80)
	/go/src/github.com/weaveworks/weave/prog/weave-npc/main.go:206 +0x4a
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(0xc4214c6600, 0xc4214c6640, 0xc4214c6620, 0x163c220, 0xc421518a80)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:168 +0x49
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc4214c6680, 0x163c220, 0xc421518a80)
	<autogenerated>:50 +0x73
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.NewInformer.func1(0x14e68c0, 0xc421534f00, 0x14e68c0, 0xc421534f00)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:246 +0x24d
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc420087970, 0xc4214c8f60, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420 +0x218
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.(*Controller).processLoop(0xc420305420)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:126 +0x3c
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.(*Controller).(github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.processLoop)-fm()
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x2a
github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil.func1(0xc4200207b0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:87 +0x5e
github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil(0xc421549fb0, 0x3b9aca00, 0x0, 0x1443901, 0xc420055bc0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:88 +0xad
github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait.Until(0xc4200207b0, 0x3b9aca00, 0xc420055bc0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/pkg/util/wait/wait.go:49 +0x4d
github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache.(*Controller).Run(0xc420305420, 0xc420055bc0)
	/go/src/github.com/weaveworks/weave/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x1c7
created by main.root
	/go/src/github.com/weaveworks/weave/prog/weave-npc/main.go:225 +0xb6a
```

**Comment by rade** [comment]
Looks like that code added to `npc.newRuleSpec` in #3064 is missing some nil checks; like in the lines just preceding it. cc @mikebryant

**Comment by bboreham** [comment]
What are the conditions which trigger this failure?

**Comment by mikebryant** [comment]
It looks like: having a `NetworkPolicy` defined that has no `from` section

**Comment by bboreham** [comment]
OK, I'm going to try extending the smoke-test (`test/840_weave_kube_3_test.sh`) with such a policy

**Comment by bboreham** [comment]
@bismarck thanks for reporting this; should be fixed in 2.0.3 now
